File: pwa-demo/src/data/repository/MediaRepository.ts
/**
 * Media Repository
 * 
 * Repository for media item and metadata operations.
 * Migrated from MediaRepository.kt
 */

import { db } from '../../services/database-complete';
import type { MediaItem, MetadataCommon } from '../local/entity';

export class MediaRepository {
  /**
   * Get media items by library
   */
  async getMediaItemsByLibrary(libraryId: number): Promise<MediaItem[]> {
    return db.mediaItems
      .where('libraryId')
      .equals(libraryId)
      .toArray();
  }

  /**
   * Get media item by ID
   */
  async getMediaItemById(itemId: number): Promise<MediaItem | undefined> {
    return db.mediaItems.get(itemId);
  }

  /**
   * Get media item by file path
   */
  async getMediaItemByPath(filePath: string): Promise<MediaItem | undefined> {
    return db.mediaItems.where('filePath').equals(filePath).first();
  }

  /**
   * Get media items by type
   */
  async getMediaItemsByType(mediaType: string): Promise<MediaItem[]> {
    return db.mediaItems
      .where('mediaType')
      .equals(mediaType)
      .toArray();
  }

  /**
   * Create new media item
   */
  async createMediaItem(mediaItem: Omit<MediaItem, 'itemId'>): Promise<number> {
    const newItem: MediaItem = {
      ...mediaItem,
      itemId: 0, // Will be auto-generated
      dateAdded: Date.now(),
      lastScanned: Date.now(),
      isAvailable: true,
      hasMetadata: false,
      hasThumbnail: false,
    };
    return db.mediaItems.add(newItem);
  }

  /**
   * Update media item
   */
  async updateMediaItem(mediaItem: MediaItem): Promise<void> {
    await db.mediaItems.update(mediaItem.itemId, {
      ...mediaItem,
      lastScanned: Date.now(),
    });
  }

  /**
   * Delete media item
   */
  async deleteMediaItem(mediaItem: MediaItem): Promise<void> {
    await db.mediaItems.delete(mediaItem.itemId);
  }

  /**
   * Search media items
   */
  async searchMediaItems(query: string, limit: number = 50): Promise<MediaItem[]> {
    const lowerQuery = query.toLowerCase();
    return db.mediaItems
      .filter(item => 
        item.fileName.toLowerCase().includes(lowerQuery) ||
        item.filePath.toLowerCase().includes(lowerQuery)
      )
      .limit(limit)
      .toArray();
  }

  /**
   * Get common metadata for item
   */
  async getCommonMetadata(itemId: number): Promise<MetadataCommon | undefined> {
    return db.metadataCommon.get(itemId);
  }

  /**
   * Save common metadata
   */
  async saveCommonMetadata(metadata: MetadataCommon): Promise<void> {
    await db.metadataCommon.put(metadata);
  }

  /**
   * Update common metadata fields
   */
  async updateCommonMetadata(
    itemId: number,
    title: string,
    summary?: string
  ): Promise<void> {
    await db.metadataCommon.update(itemId, {
      title,
      summary,
      lastUpdated: Date.now(),
    });
  }

  /**
   * Search metadata by title
   */
  async searchByTitle(query: string, limit: number = 50): Promise<MetadataCommon[]> {
    const lowerQuery = query.toLowerCase();
    return db.metadataCommon
      .filter(meta => 
        meta.title.toLowerCase().includes(lowerQuery)
      )
      .limit(limit)
      .toArray();
  }

  /**
   * Create multiple media items
   */
  async createMediaItems(mediaItems: Array<Omit<MediaItem, 'itemId'>>): Promise<number[]> {
    const items: MediaItem[] = mediaItems.map(item => ({
      ...item,
      itemId: 0,
      dateAdded: Date.now(),
      lastScanned: Date.now(),
      isAvailable: true,
      hasMetadata: false,
      hasThumbnail: false,
    }));
    
    return db.mediaItems.bulkAdd(items, { allKeys: true }) as Promise<number[]>;
  }

  /**
   * Get media item with its metadata
   */
  async getMediaItemWithMetadata(
    itemId: number
  ): Promise<{ mediaItem?: MediaItem; metadata?: MetadataCommon }> {
    const [mediaItem, metadata] = await Promise.all([
      this.getMediaItemById(itemId),
      this.getCommonMetadata(itemId),
    ]);

    return { mediaItem, metadata };
  }

  /**
   * Get item count by library
   */
  async getItemCountByLibrary(libraryId: number): Promise<number> {
    return db.mediaItems.where('libraryId').equals(libraryId).count();
  }

  /**
   * Get item count by type
   */
  async getItemCountByType(mediaType: string): Promise<number> {
    return db.mediaItems.where('mediaType').equals(mediaType).count();
  }

  /**
   * Get recent media items
   */
  async getRecentMediaItems(limit: number = 20): Promise<MediaItem[]> {
    return db.mediaItems
      .orderBy('dateAdded')
      .reverse()
      .limit(limit)
      .toArray();
  }

  /**
   * Get items without metadata
   */
  async getItemsWithoutMetadata(libraryId?: number): Promise<MediaItem[]> {
    let query = db.mediaItems.where('hasMetadata').equals(0);
    
    if (libraryId !== undefined) {
      const items = await query.toArray();
      return items.filter(item => item.libraryId === libraryId);
    }
    
    return query.toArray();
  }

  /**
   * Mark item as having metadata
   */
  async markHasMetadata(itemId: number, hasMetadata: boolean = true): Promise<void> {
    await db.mediaItems.update(itemId, { hasMetadata });
  }

  /**
   * Update thumbnail path
   */
  async updateThumbnail(itemId: number, thumbnailPath: string): Promise<void> {
    await db.mediaItems.update(itemId, {
      thumbnailPath,
      hasThumbnail: true,
    });
  }
}

// Export singleton instance
export const mediaRepository = new MediaRepository();

File: CleverFerret/src/main/java/com/universalmedialibrary/data/repository/MediaRepository.kt
package com.universalmedialibrary.data.repository

import com.universalmedialibrary.data.local.dao.MediaItemDao
import com.universalmedialibrary.data.local.dao.MetadataDao
import com.universalmedialibrary.data.local.entity.MediaItem
import com.universalmedialibrary.data.local.entity.MetadataCommon
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Repository for media item and metadata operations
 */
@Singleton
class MediaRepository @Inject constructor(
    private val mediaItemDao: MediaItemDao,
    private val metadataDao: MetadataDao
) {

    // Media Item operations
    fun getMediaItemsByLibrary(libraryId: Long): Flow<List<MediaItem>> =
        mediaItemDao.getMediaItemsByLibrary(libraryId)

    suspend fun getMediaItemById(itemId: Long): MediaItem? =
        mediaItemDao.getMediaItemById(itemId)

    suspend fun getMediaItemByPath(filePath: String): MediaItem? =
        mediaItemDao.getMediaItemByPath(filePath)

    fun getMediaItemsByType(mediaType: String): Flow<List<MediaItem>> =
        mediaItemDao.getMediaItemsByType(mediaType)

    fun getFavoriteMediaItems(): Flow<List<MediaItem>> =
        mediaItemDao.getFavoriteMediaItems()

    suspend fun createMediaItem(mediaItem: MediaItem): Long =
        mediaItemDao.insertMediaItem(mediaItem)

    suspend fun updateMediaItem(mediaItem: MediaItem) =
        mediaItemDao.updateMediaItem(mediaItem)

    suspend fun deleteMediaItem(mediaItem: MediaItem) =
        mediaItemDao.deleteMediaItem(mediaItem)

    suspend fun searchMediaItems(query: String, limit: Int = 50): List<MediaItem> =
        mediaItemDao.searchMediaItems(query, limit)

    // Metadata operations
    suspend fun getCommonMetadata(itemId: Long): MetadataCommon? =
        metadataDao.getCommonMetadata(itemId)

    suspend fun saveCommonMetadata(metadata: MetadataCommon) =
        metadataDao.insertCommonMetadata(metadata)

    suspend fun updateCommonMetadata(metadata: MetadataCommon) =
        metadataDao.updateCommonMetadata(metadata.itemId, metadata.title, metadata.summary)

    suspend fun searchByTitle(query: String, limit: Int = 50): List<MetadataCommon> =
        metadataDao.searchByTitle(query)

    suspend fun setFavorite(itemId: Long, isFavorite: Boolean) {
        mediaItemDao.setFavorite(itemId, isFavorite)
        metadataDao.setFavorite(itemId, isFavorite)
    }

    suspend fun toggleFavorite(itemId: Long): Boolean {
        val current = mediaItemDao.getMediaItemById(itemId)?.isFavorite ?: false
        val newState = !current
        setFavorite(itemId, newState)
        return newState
    }

    suspend fun createMediaItems(mediaItems: List<MediaItem>): List<Long> {
        mediaItemDao.insertMediaItems(mediaItems)
        return mediaItems.map { it.itemId } // Return the IDs
    }

    suspend fun getMediaItemWithMetadata(itemId: Long): Pair<MediaItem?, MetadataCommon?> {
        val mediaItem = getMediaItemById(itemId)
        val metadata = getCommonMetadata(itemId)
        return Pair(mediaItem, metadata)
    }

    suspend fun getItemCountByLibrary(libraryId: Long): Int =
        mediaItemDao.getItemCountByLibrary(libraryId)

    suspend fun getItemCountByType(mediaType: String): Int =
        mediaItemDao.getItemCountByType(mediaType)
    
    /**
     * PERFORMANCE OPTIMIZATION: Batch fetch metadata for multiple items at once
     * to avoid N+1 query problems.
     * 
     * @param itemIds List of item IDs to fetch metadata for
     * @return Map of itemId to MetadataCommon
     */
    suspend fun getCommonMetadataBatch(itemIds: List<Long>): Map<Long, MetadataCommon> {
        if (itemIds.isEmpty()) return emptyMap()
        return metadataDao.getMetadataCommonBatch(itemIds).associateBy { it.itemId }
    }
}
